subprojects/docs/src/docs/userguide/lazyConfiguration.adoc

```diff
@@ -15,40 +15,226 @@
+
+To make this easier, Gradle offers convenient APIs for defining files or directories as task inputs and outputs in a descriptive way. As an example consider the following plugin with a producer and consumer task, which are wired together via inputs and outputs:
+
+++++
+<sample id="implicitTaskDependency" dir="providers/implicitTaskDependency" title="Implicit task dependency">
+    <sourcefile file="build.gradle"/>
+    <output args="consumer"/>
+</sample>
+++++
+
+In the example above, the task output and input are connected before any location is defined. This is possible because the input and output properties use the `Provider` API. The output property is created with api:org.gradle.api.DefaultTask#newOutputFile()[] and the input property is created with api:org.gradle.api.DefaultTask#newInputFile()[]. Values are only resolved when they are needed during execution. The setters can be called at any time before the task is executed and the change will automatically affect all related input and output properties.
+
+Another thing to note is the absence of any explicit task dependency. Properties created via `newOutputFile()` and `newOutputDirectory()` bring knowledge about which task is generating them, so using them as task input will implicitly link tasks together.
+
+[[sec:working_with_collection]]
+=== Working with collection Providers
+
+In this section, we are going to explore lazy collections. They work exactly like any other `Provider` and, just like `FileSystemLocation` providers, they have additional modeling around them. They are implemented as a strongly typed model called api:org.gradle.api.provider.ListProperty[]. You can create a new `ListProperty` using api:org.gradle.api.model.ObjectFactory#listProperty(java.lang.Class)[] and specifying the element's type.
+
+This type of property allows you to overwrite the entire list with api:org.gradle.api.provider.Property#set(org.gradle.api.provider.Provider)[] or add new elements through the various `add` methods:
+
+* api:org.gradle.api.provider.ListProperty#add(T)[]: Add a single concrete element to the list
+* api:org.gradle.api.provider.ListProperty#add(org.gradle.api.provider.Provider)[]: Add a lazily evaluated element to the list
+* api:org.gradle.api.provider.ListProperty#addAll(org.gradle.api.provider.Provider)[]: Add a lazily evaluated list of elements to the list
+
+Just like every `Provider`, the list is evaluated when api:org.gradle.api.provider.Provider#get()[] is called. The following example show the api:org.gradle.api.provider.ListProperty[] in action:
+
+++++
+<sample id="listProperty" dir="providers/listProperty" title="List property">
+    <sourcefile file="build.gradle"/>
+    <output args="print"/>
+</sample>
+++++
 
 [[sec:lazy_configuration_faqs]]
 === FAQs
```

Comment:
I am usually not a big fan of FAQs. They indicate that the content wasn't explain good enough and still leaves open questions. Maybe we can move all that into the text above?

Reply:
Agreed. Even better if we can point to a real, annotated sample.

Reply:
As a counter point, we've had the conversation about how to use Provider/Property several times in long form and this is the summary that came out of that. Maybe we shouldn't call it FAQs, but "Recommendations and Anti-patterns".  I think in this particular case, the question and answer format cuts down on the amount of information someone has to digest.  People will ask "Why isn't there a setter?" or "Why do I have to call `.get()`?"

Reply:
I converted the section into `Guidelines`.